src/lib.rs

```diff
@@ -553,231 +561,45 @@ impl State {
-                }
-                return false;
+        loop {
+            // Try to get a runnable from the local queue.
+            if let Some(runnable) = local_ticker(false) {
+                return runnable;
             }
-        }
 
-        self.wakers.push((id, waker.clone()));
-        true
-    }
+            // Register a local waiter.
+            listener!(local.waiters => local_listener);
 
-    /// Removes a previously inserted sleeping ticker.
-    ///
-    /// Returns `true` if the ticker was notified.
-    fn remove(&mut self, id: usize) -> bool {
-        self.count -= 1;
-        self.free_ids.push(id);
-
-        for i in (0..self.wakers.len()).rev() {
-            if self.wakers[i].0 == id {
-                self.wakers.remove(i);
-                return false;
+            // Try for a global runner.
+            if let Ok(runnable) = self.queue.pop() {
+                return runnable;
             }
-        }
-        true
-    }
 
-    /// Returns `true` if a sleeping ticker is notified or no tickers are sleeping.
-    fn is_notified(&self) -> bool {
-        self.count == 0 || self.count > self.wakers.len()
-    }
-
-    /// Returns notification waker for a sleeping ticker.
-    ///
-    /// If a ticker was notified already or there are no tickers, `None` will be returned.
-    fn notify(&mut self) -> Option<Waker> {
-        if self.wakers.len() == self.count {
-            self.wakers.pop().map(|item| item.1)
-        } else {
-            None
-        }
-    }
-}
-
-/// Runs task one by one.
-struct Ticker<'a> {
-    /// The executor state.
-    state: &'a State,
-
-    /// Set to a non-zero sleeper ID when in sleeping state.
-    ///
-    /// States a ticker can be in:
-    /// 1) Woken.
-    /// 2a) Sleeping and unnotified.
-    /// 2b) Sleeping and notified.
-    sleeping: usize,
-}
+            // Register a global waiter.
+            listener!(self.new_tasks => global_listener);
 
-impl Ticker<'_> {
-    /// Creates a ticker.
-    fn new(state: &State) -> Ticker<'_> {
-        Ticker { state, sleeping: 0 }
-    }
-
-    /// Moves the ticker into sleeping and unnotified state.
-    ///
-    /// Returns `false` if the ticker was already sleeping and unnotified.
-    fn sleep(&mut self, waker: &Waker) -> bool {
-        self.state
-            .local_queue
-            .get_or_default()
-            .waker
-            .register(waker);
-
-        let mut sleepers = self.state.sleepers.lock().unwrap();
-
-        match self.sleeping {
-            // Move to sleeping state.
-            0 => {
-                self.sleeping = sleepers.insert(waker);
-            }
-
-            // Already sleeping, check if notified.
-            id => {
-                if !sleepers.update(id, waker) {
-                    return false;
-                }
+            // Try for both again.
+            if let Some(runnable) = local_ticker(true).or_else(|| self.queue.pop().ok()) {
```

Comment:
Do we try to pull from `self.queue.pop` twice simultaneously here usually?

Reply:
No, `local_ticker` only reads from local queues.